src/MOI/MOI_wrapper.jl

```diff
@@ -2738,7 +2768,82 @@ function MOI.optimize!(model::Optimizer)
         _update_MIP_start!(model)
     end
     start_time = time()
-    if is_mip(model)
+    if is_nlp(model)
+        ncols=n_variables(model.inner)
+        # Getting problem's variables
+        x=collect(keys(model.variable_info))
+        # Creating a NULL Objective function to allow its modification by XPRSnlp
+        c=[0.0 for i = 1:ncols]
+        MOI.set(model,
+        MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}(),
+        MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(c, x), 0.0),
+        );
+        # Changing names to the right format adapted to XPRSnlp
+        names=["x$idx" for idx = 1:ncols]
+        MOI.set(model, MOI.VariableName(), x, names)  
+        Xpress._pass_variable_names_to_solver(model)  
+        
+        count=length(model.affine_constraint_info)
+        #Case with NL Objective and no NL Constraints
+        if length(model.nlp_constraint_info)==0 
+            # Delete existing constraints when optimize! is called again
+            if count>0 
+                for i in 1:count
+                    pop!(model.affine_constraint_info,collect(keys(model.affine_constraint_info))[1])
+                end
+            end
+
+        #Case with NL Constraints
+        
+            #New NL Constraints
+        elseif count != length(model.nlp_constraint_info)
+        # Delete existing constraints when optimize! is called again
+            for i in 1:count
+                popfirst!(model.nlp_constraint_info)
+                pop!(model.affine_constraint_info,collect(keys(model.affine_constraint_info))[1])
+            end
+        
+        
+            # Creating NULL constraints to allow their modification by XPRSnlp
+            for cons in model.nlp_constraint_info
+                c_set=to_constraint_set(cons)
+                if length(c_set)==2
+                    MOI.add_constraint(
+                    model,
+                    MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(c, x), 0.0),
+                    c_set[1],
+                    );
+
+                    MOI.add_constraint(
+                    model,
+                    MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(c, x), 0.0),
+                    c_set[2],
+                    );
+                elseif c_set !== nothing
+                    MOI.add_constraint(
+                    model,
+                    MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(c, x), 0.0),
+                    c_set[1],
+                    );
+                end
+            end
+        
+
+            # Passing constraints to the solver
+            for i in 0:length(model.nlp_constraint_info)-1
+                Lib.XPRSnlpchgformulastring(model.inner, Cint(i), join(["+"," ",to_str(model.nlp_constraint_info[i+1].expression)]))
```

Comment:
For realz? The API is to pass it as a string?

Reply:
Ah. Seems like you can also pass tape:
https://www.fico.com/fico-xpress-optimization/docs/latest/solver/nonlinear/HTML/chapAPIFuncs.html?scroll=secAddNLPart

Reply:
yes, there is a tape. we were lazy though.